### PR TITLE
Begin splitting middle::ty

### PR DESCRIPTION
**Pull request by arielb1**
That file got way too big for its own good. It could be split more - this is just a start.

r? @nikomatsakis

### CODE REVIEW

**Comment by arielb1** [comment]
Let the bikeshedding commence.

**Comment by nikomatsakis** [comment]
I basically like this. I'm not sure about moving all the impls to a single module, but it might be nice. I've been wanting to split ty.rs into a directory for some time, for sure.

@arielb1 this is just a reorganization, right? No behavioral changes?

**Comment by arielb1** [comment]
@nikomatsakis 

Right. No behavioral or (intentional) performance changes.

**Comment by arielb1** [comment]
I didn't move *all* impls, just `Lift`, `HasTypeFlags`, `RegionEscape` and `TypeFoldable`.

File: src/librustc/middle/ty/ivar.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+/// it can be referred to from Ty) but can only be modified if its
+/// lifetime is exactly 'tcx.
+///
+/// Safety invariants:
+///     (A) self.0, if fulfilled, is a valid Ty<'tcx>
+///     (B) no aliases to this value with a 'tcx longer than this
+///         value's 'lt exist
+///
+/// NonZero is used rather than Unique because Unique isn't Copy.
+pub struct TyIVar<'tcx, 'lt: 'tcx>(ivar::Ivar<NonZero<*const TyS<'static>>>,
```
Comment (jroesch) [inline src/librustc/middle/ty/ivar.rs]: I'm not a huge fan of having a redundant prefix on the name, but I guess you are doing it to disambiguate with Ivar.

**Comment by jroesch** [comment]
I like this as well :+1: after a 1214, MIR, and HIR this is going to be another "fun" rebase on my branch.

File: src/librustc/middle/ty/ivar.rs (hunk 1 above)
Comment (arielb1) [inline src/librustc/middle/ty/ivar.rs]: It is an `Ivar<Ty<'tcx>>`, not inline because of variance being annoying.

**Comment by nikomatsakis** [comment]
@arielb1 

> I didn't move all impls, just Lift, HasTypeFlags, RegionEscape and TypeFoldable.

So I guess the interesting question is where these impls should go in general. I've sometimes regretted the current style -- I tend to think the impls should probably go with the types (though it's natural when first defining the traits to stick them with the trait). This is particularly true since in the cross-crate case they *must* go with the types. Note that "impls go with the types" might be loosely defined -- e.g., all types defined in the `ty` module might put their impls under `ty::impls`, but things in `traits` might be under `traits::impls`, etc.

**Comment by nikomatsakis** [comment]
r+ from me, in any case. This looks like progress to me. I second the... joyful thoughts of rebases to come.

**Comment by arielb1** [comment]
@nikomatsakis 

90% of the impls are for ty and rustc dependencies (std, syntax, etc.), and therefore would go in `ty::structural_impls` anyway. Should I create a similar `traits::structural_impls` for these?

**Comment by nikomatsakis** [comment]
@arielb1 I think I would prefer that, yes. The `Debug` impls for `traits` are already stashed in `util.rs`, they could be combined together as well (basically, all the... de rigeur impls).

**Comment by nikomatsakis** [comment]
But then I see you already did it.

**Comment by munyari** [comment]
Are `InfrigingField` and `InfrigingVariant` not typos?

**Pull request merged**